### ISSUE TITLE
Rule proposal: `no-this-assignment`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#997]
Forbid assign `this` to another variable, I don't see any reason to do this.

## Fail

```js
const that = this;
function foo() {
	return that.foo;
}
foo();
```

```js
const that = this;
class Foo {
	bar() {
		console.log(that.foo)
	}
}
new Foo().bar();
```

## Pass

```js
const foo = () => this.foo;
foo();
```

```js
function foo(someObject) {
	return someObject.foo;
}
foo(this);
```

```js
class Foo {
	constructor(someObject) {
		this.someObject = someObject
	}
	bar() {
		console.log(this.someObject.foo)
	}
}

new Foo(this).bar()
```

I guess we should allow this case

```js
const foo = {someObject: this};
```

Because we don't know how this `foo` is used.
